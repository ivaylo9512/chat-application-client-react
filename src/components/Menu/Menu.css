.circle-nav nav {
    position: relative;
    height: 32vw;
    width: 32vw;
}
.circle-nav li {
    position: absolute;
    font-size: 1.5em;
    width: 4vw;
    height: 100%;
    left: 50%;
    margin-left: -2vw;
    transform-origin: center;
    list-style: none;
}
.circle-nav li a {
    text-align: center;
    line-height: 3vw;
}
.circle-nav li button,
.circle-nav li a {
    position: absolute;
    top: 1vw;
    border-radius: 50%;
    font-size: 0.95vw;
    color: white;
    height: 3vw;
    width: 3vw;
    background: rgb(24,27, 33);
    padding: 0;
    border: 0;
    box-shadow: 0px 0px 0.5vw 0.1vw rgb(0, 0, 0, 0.8)
} 
.circle-nav li a span {
    font-size: 1.1em;
    opacity: 0.7;
}
.circle-nav li:first-child {
    transform: rotate(-5deg);
}
.circle-nav li:nth-child(2) {
    transform: rotate(15deg);
}
.circle-nav li:nth-child(3) {
    transform: rotate(35deg);
}
.circle-nav li:nth-child(4) {
    transform: rotate(55deg);
}
.circle-nav li:nth-child(5) {
    transform: rotate(75deg);
}
.circle-nav li:nth-child(6) {
    transform: rotate(95deg);
}
.circle-nav li:nth-child(6) {
    transform: rotate(95deg);
}
.circle-nav li:nth-child(6) {
    transform: rotate(95deg);
}
.circle-nav li:nth-child(7) {
    transform: rotate(120deg);
}
.circle-nav li:nth-child(8) {
    transform: rotate(140deg);
}
.circle-nav li:nth-child(9) {
    transform: rotate(160deg);
}