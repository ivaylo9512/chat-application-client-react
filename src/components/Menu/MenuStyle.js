import styled from 'styled-components'

export const MenuCircle = styled.button`
    border-radius: 50%;
    border: 0px;
    box-shadow: 0px 0px 3px 1px rgba(0,0,0,0.75);
    height: 2.7em;
    width: 2.7em;
    background: rgb(24,27, 33);
    color: white;
    font-size: 1.2em;
    position: absolute;
    top: 100%;
    left: -2em;
`

export const MenuContainer = styled.div`
    width: ${props => props.isHidden 
        ? '0%' 
        : '27%'
    };
    transition: all 4s;

    ${MenuCircle}{
        transition: ${props => props.isHidden 
            ? 'all 2s 1.75s' 
            : 'all 2s'
        };
        transform: ${props => props.isHidden 
            ? 'translate(4em, -5em)'
            : 'translate(0, 0)'
        };
    }

    > div{
        height: 100%;
        width: 100%;
        transform: ${props => props.isHidden 
            ? 'translate(-22.5em, 18em);'
            : 'translate(0, 0)'
        };
        transition: all 4s; 
    }
`

export const CircleNav = styled.div`
    overflow: hidden;
    position: fixed;
    bottom: -11.6vw;
    border: 0.5px solid #c2001b;
    right: 100%;
    border-radius: 50%;
    margin-right: -26.8vw;
    background: rgb(24,27, 33);
    display: flex;
    align-items: center;
    justify-content: center;
    box-shadow: 0px 0px 0.4vw 0.2vw rgb(0, 0, 0, 0.75);


    ul{
        margin: 0;
    }
`

export const CircleBtn = styled.button`
    background: rgb(24,27, 33);
    border: 0px;
    color: #c2001b;
    font-size: 2.3vw;
    height: 22.5vw;
    width: 22.5vw;
    position: absolute;
    right: 50%;
    top: 50%;
    margin-right: -11.25vw;
    margin-top: -11.25vw;
    bottom: -6.7vw;
    border-radius: 50%;
    cursor: pointer;
    z-index: 3;
    padding: 0;
`