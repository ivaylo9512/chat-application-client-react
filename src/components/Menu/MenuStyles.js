import styled from 'styled-components'

export const MenuCircle = styled.button`
    border-radius: 50%;
    border: 0px;
    box-shadow: 0px 0px 0.2vw 0.1vw rgba(0,0,0,0.75);
    height: 4vw;
    width: 4vw;
    background: rgb(24,27, 33);
    color: white;
    font-size: 1.4vw;
    position: absolute;
    bottom: -4vw;
    left: -4vw;

    transition: ${props => props.isHidden 
        ? 'transform 2s 1.75s' 
        : 'transform 2s'
    };
    transform: ${props => props.isHidden 
        ? 'translate(7vw, -7vw)'
        : 'translate(0, 0)'
    };
`

export const MenuContainer = styled.div`
    width: ${props => props.isHidden 
        ? '0%' 
        : '27%'
    };
    transition: all 4s;

    > div{
        height: 100%;
        width: 100%;
        transform: ${props => props.isHidden 
            ? 'translate(-27.8vw, 21vw);'
            : 'translate(0, 0)'
        };
        transition: all 4s; 
    }
`

export const CircleNav = styled.div`
    overflow: hidden;
    position: fixed;
    bottom: -13.2vw;
    margin-right: -19.8vw;
    border: 0.1vw solid white;
    right: 100%;
    border-radius: 50%;
    background: rgb(24,27, 33);
    display: flex;
    align-items: center;
    justify-content: center;
    box-shadow: 0px 0px 0.3vw 0.1vw rgb(0 0 0 / 65%);


    ul{
        margin: 0;
    }
`

export const CircleBtn = styled.button`
    background: rgb(24,27, 33);
    border: 0px;
    color: white;
    font-size: 2.3vw;
    height: 22.5vw;
    width: 22.5vw;
    position: absolute;
    right: 50%;
    top: 50%;
    margin-right: -11.25vw;
    margin-top: -11.25vw;
    bottom: -6.7vw;
    border-radius: 50%;
    cursor: pointer;
    z-index: 3;
    padding: 0;
`