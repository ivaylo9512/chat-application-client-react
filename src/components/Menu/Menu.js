import { useRef, useState } from 'react'
import MenuNav from 'components/MenuNav/MenuNav'
import { MenuContainer, MenuCircle, CircleNav, CircleBtn } from './MenuStyles';

const Menu = () => {
    const [rotate, setRotate] = useState({});
    const [isMenuHidden, setIsMenuHidden] = useState(true)
    const [rotationStyle, setrotationStyle] = useState(0)
    const isRotating = useRef(false)
    const rotatingSign = useRef()
    
    const rotateNav = (e) => {
        const sign = Math.sign(e.deltaY + e.deltaX)
        if(!isRotating.current || sign != rotatingSign.current){
            rotatingSign.current = sign
            const newDeg = 20 * sign
            const max = -rotationStyle
            const min = -30 - rotationStyle  
            smoothRotate(Math.min(Math.max(newDeg, min), max), 650, rotationStyle)
        }
    }

    const smoothRotate = (deg, durration, startPos) => {
        let startTime;

        const rotate = (currentTime) => {
            isRotating.current = true
            if(!startTime) startTime = currentTime;
            const timeElapsed = currentTime - startTime;

            const amount = linearTween(timeElapsed, startPos, deg, durration);
            const rotateString = `rotate(${amount}deg)`
            setRotate({WebkitTransform: rotateString, MozTransform: rotateString, OTransform: rotateString, transform: rotateString})
            setrotationStyle(amount)
            
            if(durration / 2 < timeElapsed ) isRotating.current = false
            if(timeElapsed < durration) return requestAnimationFrame(rotate);
        }

        const linearTween = (t, b, c, d) => c*t/d + b;

        requestAnimationFrame(rotate);
    }

    return (
        <MenuContainer isHidden={isMenuHidden}>
            <MenuCircle onClick={() => setIsMenuHidden(!isMenuHidden)} tabIndex='-1'><i className='fas fa-bars'></i></MenuCircle>
            <div>
                <CircleNav onWheel={rotateNav}>
                    <CircleBtn onClick={() => setIsMenuHidden(!isMenuHidden)} tabIndex='-1'>-</CircleBtn>
                    <MenuNav rotate={rotate}/>
                </CircleNav>
            </div>
        </MenuContainer>
    )
}

export default Menu