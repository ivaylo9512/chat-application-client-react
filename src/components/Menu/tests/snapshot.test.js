import { shallow } from 'enzyme';
import Menu from 'components/Menu/Menu';
import { MenuCircle } from 'components/Menu/MenuStyles';

describe('Menu snapshot tests', () => {
    let wrapper;
    
    beforeEach(() => {
        wrapper = shallow(
            <Menu />
        )
    })
    
    it('should match snapshot with isMenuHidden to true', () => {
        expect(wrapper).toMatchSnapshot();
    })

    it('should match snapshot with isMenuHidden to false', () => {
        wrapper.find(MenuCircle).simulate('click');
        expect(wrapper).toMatchSnapshot();
    })

})