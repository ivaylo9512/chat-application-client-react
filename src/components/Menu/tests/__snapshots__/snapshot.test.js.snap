// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Menu snapshot tests should match snapshot with isMenuHidden to false 1`] = `
<styled.div
  isHidden={false}
>
  <styled.button
    isHidden={false}
    onClick={[Function]}
    tabIndex="-1"
  >
    <i
      className="fas fa-bars"
    />
  </styled.button>
  <div>
    <styled.div
      onWheel={[Function]}
    >
      <styled.button
        onClick={[Function]}
        tabIndex="-1"
      >
        -
      </styled.button>
      <MenuNav
        rotate={Object {}}
      />
    </styled.div>
  </div>
</styled.div>
`;

exports[`Menu snapshot tests should match snapshot with isMenuHidden to true 1`] = `
<styled.div
  isHidden={true}
>
  <styled.button
    isHidden={true}
    onClick={[Function]}
    tabIndex="-1"
  >
    <i
      className="fas fa-bars"
    />
  </styled.button>
  <div>
    <styled.div
      onWheel={[Function]}
    >
      <styled.button
        onClick={[Function]}
        tabIndex="-1"
      >
        -
      </styled.button>
      <MenuNav
        rotate={Object {}}
      />
    </styled.div>
  </div>
</styled.div>
`;
